### ISSUE TITLE
`import defer` testing plan

### ISSUE THREAD

**Issue by nicolo-ribaudo** [tc39/test262#4215]
https://github.com/tc39/proposal-defer-import-eval/

### Valid cases

```js
import defer * as ns from "mod";
```
```js
import defer from "mod";
```
```js
import defer * as ns from "mod";
export { ns };
```

### Invalid cases (one per file)

```js
import defer x from "mod";
import defer { x } from "mod";
import defer y, * as ns from "mod";
import defer as ns from "mod";
export defer * as ns from "mod";
```

- [x] All the existing dynamic import syntax tests should be duplicated to `import.defer` (both valid and invalid) (https://github.com/tc39/test262/pull/4374)

## Deferred namespace object

- [x] It is an object
- [x] It is not the same object as the non-deferred one
- [x] Importing twice gives the same object
    - [x] static/static
    - [x] static/dynamic
    - [x] dynamic/dynamic
    - [x] static from different files
- [x] import and import defer + exec only execute the module once
- [x] `@@toStringTag` (https://github.com/tc39/test262/pull/4360)

### Sync

- [x] `import defer` does not trigger evaluation
- [x] `import.defer` does not trigger evaluation (https://github.com/tc39/test262/pull/4375)
- [x] `[[Get]]` triggers evaluation
    - [x] with a known export name
    - [x] with a non-exported name
    - [x] does not trigger evaluation with a symbol
- [x] What operations trigger evaluation (https://github.com/tc39/test262/pull/4341)

### Errors

- [x] Property access re-throws the evaluation error
    - [x] On first access for not evaluated modules
    - [x] On second access for not evaluated modules
    - [x] On first access for already evaluated modules
- [x] Import defer of an errored module doesn't throw
- [x] `ReadyForSyncExecution`
    - [x] Cannot evaluate a module in its _evaluating_ state
    - [x] Cannot evaluate itself
    - [x] Cannot evaluate a module with a dependency in an _evaluating_ state
    - [x] The error when evaluating a module with two dependencies, where the first one is _linked_ and the second one is _evaluating_, happens before evaluating the first one
    - [x] A deferred module evaluates itself while being deferred-evaluated
    - [x] Cannot evaluate a module in its _evaluating-async_ state
    - [x] Cannot evaluate itself after a TLA
    - [x] Cannot evaluate a module with a dependency in an _evaluating-async_ state.
- [x] Syntax error are not deferred
- [x] Linking errors are not deferred

### Async

- [x] Async modules are not deferred
- [x] Sync modules with async dep are deferred, but their async deps are not
- [x] Sync deps of async modules are not deferred
- [x] Async transitive deps are flattened and exexuted in the right order
- [x] All these tests also for dynamic `import.defer()` (https://github.com/tc39/test262/pull/4375)

**Comment by ptomato** [comment]
@nicolo-ribaudo Can you suggest someone from whom it would be useful to get feedback on this testing plan?

**Comment by nicolo-ribaudo** [comment]
Maybe @takikawa who's working on an implementation, or @guybedford.

**Comment by linusg** [comment]
Under "valid cases" you list this example:

```js
import defer from "mod";
```

Just to clarify - that's still a regular import of an `export default` creating a binding named `defer` and not actually changed by this proposal, right?

**Comment by nicolo-ribaudo** [comment]
Yes, the test is checking that the behavior is the same as it was in ES2015, and that implementations don't accidentally mess up parsing.
